content/generators/sudoku.md

```diff
@@ -0,0 +1,163 @@
+
+
+```{code-cell} ipython3
+import networkx as nx
+import numpy as np
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+from random import sample
+```
+
+Informally the Sudoku graph is an undirected graph- its vertices represent the cells and edges represent pairs of cells that belong to the same row, column, or block of the puzzle. Formally this can be defined as:
+
+> A Sudoku grid of rank $n$ is a $n^2 × n^2$ grid($X_n$) . It consists of $n^2$ disjoint $n × n$ grids. Graph of $X_n$, denoted as $GX_n$, is $(V, E)$ where cells of Sudoku grid form the vertices of its graph and two
+cells are adjacent if they are either in the same row or column or block of $X_n$.
+
+> $GX_n$ is a regular $(n^4, \frac{3n^6}{2} − n^5− \frac{n^4}{2})$ graph of degree $3n^2 − 2n − 1$  (1)
+
+Lala, C. "Graph theory of Sudoku." (2013).
+
+Now, from (1) we can get that the graph of a Sudoku grid of rank 3 is a (81, 810) regular graph of degree 20. This can be verified informally- we have 81 cells in the standard sudoku where every cell is adjacent to 8 cells in its row + 8 cells in its column and 4 more leftover cells in its block, hence the degree 20.
+
+Let's generate a random (filled) sudoku of rank n and then build further intuition
+
+
+```{code-cell} ipython3
+def generate_random_sudoku(n):
+    rank  = n
+    side  = rank*rank
+    def pattern(row,col): return (rank*(row%rank)+row//rank+col)%side
+    def shuffle(s): return sample(s,len(s))
+    # generate random row and column indices
+    rows  = [ x*rank + row for x in shuffle(range(rank)) for row in shuffle(range(rank)) ]
+    cols  = [ x*rank + col for x in shuffle(range(rank)) for col in shuffle(range(rank)) ]
+
+    numbers  = shuffle(range(1,rank*rank+1))
+
+    board = [ [numbers[pattern(row,col)] for col in cols] for row in rows ]
+    board = np.asarray(board)
+    return board
+
+# Now we want to create the structure of a sudoku graph which we can fill up with the random grid we generated.
+# Note how we store the three different kinds of edges separately, this will be useful for visualization
+def empty_sudoku_graph(n):
+    blank_board = np.arange(n*n*n*n).reshape(n*n,n*n)
+    def get_edges(nodes):
+        edges = [(a, b) for idx, a in enumerate(nodes) for b in nodes[idx + 1:]]
+        return edges
+    box_edges=[]
+    row_edges=[]
+    column_edges=[]
+    for iteration in range(n*n):
+        i = iteration//n
+        j = iteration%n
+        column = np.asarray(blank_board)[:,iteration]
+        row = np.asarray(blank_board)[iteration,:]
+        box=blank_board[(i)*n:(i+1)*n,(j)*n:(j+1)*n].flatten()
+        box_edges += get_edges(box)
+        row_edges += get_edges(row)
+        column_edges += get_edges(column)
+    return box_edges, row_edges, column_edges
+
+# We have the edges and the nodes, now we want a networkx graph
+def draw_networkx_graph(n, board, box_edges, row_edges, column_edges, layout='grid'):
```

Comment:
I think it would be an improvement to alter the name of this function slightly, only because it is *very* similar to some of the drawing function names in `nx_pylab`. Maybe `draw_sudoku` or `visualize_sudoku_graph`?

Ultimately I think it'd also be an improvement to move some of the processing done in the function (i.e. the graph building and layout handling) outside of the function, but that will be easier to discuss in a follow-up PR IMO, so I wouldn't worry about it now!